test/e2e/uniswap_match.ts

```diff
@@ -0,0 +1,199 @@
+
+import GPv2Settlement from "../../build/artifacts/src/contracts/GPv2Settlement.sol/GPv2Settlement.json";
+import { domain } from "../../src/ts";
+import { Order, OrderKind, SigningScheme } from "../../src/ts/order";
+import { SettlementEncoder } from "../../src/ts/settlement";
+import type { SignatureLike, TypedDataDomain } from "../../src/ts/types/ethers";
+
+export interface UniswapTradeRequest {
+  sellToken: string;
+  buyToken: string;
+  sellAmount: BigNumberish;
+}
+
+export interface UniswapPairState {
+  uniswapPair: Contract;
+  sellToken: Contract;
+  reserveSellToken: BigNumber;
+  buyToken: Contract;
+  reserveBuyToken: BigNumber;
+  isSellTokenToken0: boolean;
+}
+
+async function getUniswapState(
+  uniswapFactory: Contract,
+  { sellToken, buyToken }: { sellToken: string; buyToken: string },
+  provider: providers.JsonRpcProvider,
+): Promise<UniswapPairState> {
+  const uniswapPair = new Contract(
+    await uniswapFactory.getPair(sellToken, buyToken),
+    UniswapV2Pair.abi,
+    provider,
+  );
+  const sellTokenContract = new Contract(sellToken, ERC20.abi, provider);
+  const buyTokenContract = new Contract(buyToken, ERC20.abi, provider);
+
+  const [reserveSellToken, reserveBuyToken, pairToken0] = await Promise.all([
+    sellTokenContract.balanceOf(uniswapPair.address),
+    buyTokenContract.balanceOf(uniswapPair.address),
+    uniswapPair.token0(),
+  ]);
+
+  const isSellTokenToken0 = sellToken === pairToken0;
+  console.log(isSellTokenToken0);
+  return {
+    uniswapPair,
+    reserveSellToken,
+    reserveBuyToken,
+    sellToken: sellTokenContract,
+    buyToken: buyTokenContract,
+    isSellTokenToken0,
+  };
+}
+
+/**
+ * A very basic solver that takes the amount a user wishes to trade on Uniswap
+ * and uses it to:
+ * 1. transform the swap into a GPv2 sell order to sign;
+ * 2. given the signed order, settle the trade onchain.
+ *
+ * This class is single use: each new Uniswap order requires a new instance of
+ * this class.
+ */
+export class UniswapMatch {
+  private readonly uniswapFactory: Contract;
+  private readonly settlement: Contract;
+  private readonly domainSeparator: Promise<TypedDataDomain>;
+  private readonly provider: providers.JsonRpcProvider;
+  order?: Order;
+  uniswapState?: UniswapPairState;
+
+  constructor({
+    uniswapFactoryAddress,
+    settlementAddress,
+    provider,
+  }: {
+    uniswapFactoryAddress: string;
+    settlementAddress: string;
+    provider: providers.JsonRpcProvider;
+  }) {
+    this.uniswapFactory = new Contract(
+      uniswapFactoryAddress,
+      UniswapV2Factory.abi,
+      provider,
+    );
+    this.settlement = new Contract(
+      settlementAddress,
+      GPv2Settlement.abi,
+      provider,
+    );
+    this.provider = provider;
+    this.domainSeparator = provider
+      .getNetwork()
+      .then(({ chainId }) => domain(chainId, this.settlement.address));
+  }
+
+  async createSellOrder({
```

Comment:
I think this is a nice method to have (basically allowing you to simulate the functionality of the swap interface), but I think it should be side-effect less. That is it should return an order and then I should be able to call `submitSolution` either with this order or with any other (potentially frontend generated) order.

Then we could use this tool to also clear orders that are stuck in the system (in case the solver goes down for some reason or decides to not settle a specific order).